pkg/controller/garbagecollector/garbagecollector.go

```diff
@@ -191,30 +191,82 @@ func (gc *GarbageCollector) Sync(discoveryClient discovery.ServerResourcesInterf
-		// pass a timeout channel, but we have to consider the implications of
-		// un-pausing the GC with a partially synced graph builder.
-		if !controller.WaitForCacheSync("garbage collector", stopCh, gc.dependencyGraphBuilder.IsSynced) {
-			utilruntime.HandleError(fmt.Errorf("timed out waiting for dependency graph builder sync during GC sync"))
-			return
+		// Once we get here, we should not unpause workers until we've successfully synced again
+		for attempt := 0; ; attempt++ {
+			// If this is a re-attempt, check if available resources have changed
+			if attempt > 0 {
+				newResources = GetDeletableResources(discoveryClient)
+			}
+
+			added := map[schema.GroupVersionResource]bool{}
+			removed := map[schema.GroupVersionResource]bool{}
+			for old := range oldResources {
+				if _, ok := newResources[old]; !ok {
+					removed[old] = true
+				}
+			}
+			for new := range newResources {
+				if _, ok := oldResources[new]; !ok {
+					added[new] = true
+				}
+			}
+			glog.V(2).Infof("added since last successful sync: %#v", added)
+			glog.V(2).Infof("removed since last successful sync: %#v", removed)
+
+			// Resetting the REST mapper will also invalidate the underlying discovery
+			// client. This is a leaky abstraction and assumes behavior about the REST
+			// mapper, but we'll deal with it for now.
+			gc.restMapper.Reset()
+			glog.V(2).Infof("reset restmapper")
+
+			// Perform the monitor resync and wait for controllers to report cache sync.
+			//
+			// NOTE: It's possible that newResources will diverge from the resources
+			// discovered by restMapper during the call to Reset, since they are
+			// distinct discovery clients invalidated at different times. For example,
+			// newResources may contain resources not returned in the restMapper's
+			// discovery call if the resources appeared in-between the calls. In that
+			// case, the restMapper will fail to map some of newResources until the next
+			// sync period.
+			if err := gc.resyncMonitors(newResources); err != nil {
+				utilruntime.HandleError(fmt.Errorf("failed to sync resource monitors (attempt %d): %v", attempt, err))
+				continue
+			}
+
+			glog.V(2).Infof("resynced monitors")
+
+			// wait up to our sync period before bailing, rediscovering resources, and trying again.
+			// note that workers stay paused.
+			// shared informers will keep attempting to sync in the background, so we're not interrupting them.
+			// resyncMonitors will no-op on resources that still exist on the next attempt.
+			waitForSyncStop := make(chan struct{})
+			go func() {
+				// block
+				select {
+				// until we're told to stop
+				case <-stopCh:
+				// or until we hit our timeout
+				case <-time.After(period):
+				}
+				// make sure the WaitForCacheSync call returns
+				close(waitForSyncStop)
+			}()
+
+			if controller.WaitForCacheSync("garbage collector", waitForSyncStop, gc.dependencyGraphBuilder.IsSynced) {
```

Comment:
nit: the monitor-cache-sync progress (for resources that aren't added or removed) won't get reset when we timeout, so we don't need exponential backoff timeout

the "timeout" is not limiting the time resource used for syncing cache. not sure if it's a pro or con

Reply:
> the "timeout" is not limiting the time resource used for syncing cache.

Can you clarify what you mean by this

Reply:
> nit: the monitor-cache-sync progress (for resources that aren't added or removed) won't get reset when we timeout, so we don't need exponential backoff timeout

Agree we don't need exponential, a simple timeout before we attempt rediscovery/resync seems sufficient

Reply:
E.g. we discover a new resource and start a new monitor for it with `gc.resyncMonitors()`. We set a timeout of 30s and wait for the monitor cache to get synced. A long sync time (like 50s) won't violate the timeout. We don't need to increase the timeout, instead we just wait till the second timeout period. It was confusing to me that the timeout is not for guarding the sync time, it's the period when we decide to poll discovery again.

Reply:
got it... that's what this comment was trying to explain, though I'm completely open to better wording (*I* understood what I meant, but... I didn't get that understanding from reading my own comment)

```go
// wait for caches to fill for a while (our sync period) before attempting to rediscover resources and retry syncing.
// this protects us from deadlocks where available resources changed and one of our informer caches will never fill.
// informers keep attempting to sync in the background, so retrying doesn't interrupt them.
// the call to resyncMonitors on the reattempt will no-op for resources that still exist.
// note that workers stay paused until we successfully resync.
waitForSyncStop := make(chan struct{})
```